File: onedal/ensemble/forest.py
# ==============================================================================
# Copyright 2023 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

import numbers
import warnings
from abc import ABCMeta, abstractmethod
import math

from daal4py.sklearn._utils import daal_check_version
from sklearnex import get_hyperparameters

from ..common._base import BaseEstimator
from ..common._estimator_checks import _check_is_fitted
from ..common._mixin import ClassifierMixin, RegressorMixin
from ..datatypes import _convert_to_supported, from_table, to_table


class BaseForest(BaseEstimator, metaclass=ABCMeta):
    @abstractmethod
    def __init__(
        self,
        n_estimators,
        criterion,
        max_depth,
        min_samples_split,
        min_samples_leaf,
        min_weight_fraction_leaf,
        max_features,
        max_leaf_nodes,
        min_impurity_decrease,
        min_impurity_split,
        bootstrap,
        oob_score,
        random_state,
        warm_start,
        class_weight,
        ccp_alpha,
        max_samples,
        max_bins,
        min_bin_size,
        infer_mode,
        splitter_mode,
        voting_mode,
        error_metric_mode,
        variable_importance_mode,
        algorithm,
        **kwargs,
    ):
        self.n_estimators = n_estimators
        self.bootstrap = bootstrap
        self.oob_score = oob_score
        self.random_state = random_state
        self.warm_start = warm_start
        self.class_weight = class_weight
        self.max_samples = max_samples
        self.criterion = criterion
        self.max_depth = max_depth
        self.min_samples_split = min_samples_split
        self.min_samples_leaf = min_samples_leaf
        self.min_weight_fraction_leaf = min_weight_fraction_leaf
        self.max_features = max_features
        self.max_leaf_nodes = max_leaf_nodes
        self.min_impurity_decrease = min_impurity_decrease
        self.min_impurity_split = min_impurity_split
        self.ccp_alpha = ccp_alpha
        self.max_bins = max_bins
        self.min_bin_size = min_bin_size
        self.infer_mode = infer_mode
        self.splitter_mode = splitter_mode
        self.voting_mode = voting_mode
        self.error_metric_mode = error_metric_mode
        self.variable_importance_mode = variable_importance_mode
        self.algorithm = algorithm

    def _to_absolute_max_features(self, n_features):
        if self.max_features is None:
            return n_features
        elif isinstance(self.max_features, str):
            return max(1, int(getattr(math, self.max_features)(n_features)))
        elif isinstance(self.max_features, numbers.Integral):
            return self.max_features
        elif self.max_features > 0.0:
            return max(1, int(self.max_features * n_features))
        return 0

    def _get_observations_per_tree_fraction(self, n_samples, max_samples):
        if max_samples is None:
            return 1.0

        if isinstance(max_samples, numbers.Integral):
            if not (1 <= max_samples <= n_samples):
                msg = "`max_samples` must be in range 1 to {} but got value {}"
                raise ValueError(msg.format(n_samples, max_samples))
            return max(float(max_samples / n_samples), 1 / n_samples)

        if isinstance(max_samples, numbers.Real):
            return max(float(max_samples), 1 / n_samples)

        msg = "`max_samples` should be int or float, but got type '{}'"
        raise TypeError(msg.format(type(max_samples)))

    def _get_onedal_params(self, data):
        n_samples, n_features = data.shape

        self.observations_per_tree_fraction = self._get_observations_per_tree_fraction(
            n_samples=n_samples, max_samples=self.max_samples
        )
        self.observations_per_tree_fraction = (
            self.observations_per_tree_fraction if bool(self.bootstrap) else 1.0
        )
        min_observations_in_leaf_node = (
            self.min_samples_leaf
            if isinstance(self.min_samples_leaf, numbers.Integral)
            else int(math.ceil(self.min_samples_leaf * n_samples))
        )

        min_observations_in_split_node = (
            self.min_samples_split
            if isinstance(self.min_samples_split, numbers.Integral)
            else int(math.ceil(self.min_samples_split * n_samples))
        )

        onedal_params = {
            "fptype": data.dtype,
            "method": self.algorithm,
            "infer_mode": self.infer_mode,
            "voting_mode": self.voting_mode,
            "observations_per_tree_fraction": self.observations_per_tree_fraction,
            "impurity_threshold": float(
                0.0 if self.min_impurity_split is None else self.min_impurity_split
            ),
            "min_weight_fraction_in_leaf_node": self.min_weight_fraction_leaf,
            "min_impurity_decrease_in_split_node": self.min_impurity_decrease,
            "tree_count": int(self.n_estimators),
            "features_per_node": self._to_absolute_max_features(n_features),
            "max_tree_depth": int(0 if self.max_depth is None else self.max_depth),
            "min_observations_in_leaf_node": min_observations_in_leaf_node,
            "min_observations_in_split_node": min_observations_in_split_node,
            "max_leaf_nodes": (0 if self.max_leaf_nodes is None else self.max_leaf_nodes),
            "max_bins": self.max_bins,
            "min_bin_size": self.min_bin_size,
            "seed": self.random_state,
            "memory_saving_mode": False,
            "bootstrap": bool(self.bootstrap),
            "error_metric_mode": self.error_metric_mode,
            "variable_importance_mode": self.variable_importance_mode,
        }
        if isinstance(self, ClassifierMixin):
            onedal_params["class_count"] = (
                0 if self.classes_ is None else len(self.classes_)
            )
        if daal_check_version((2023, "P", 101)):
            onedal_params["splitter_mode"] = self.splitter_mode
        return onedal_params

    def _fit(self, X, y, sample_weight, module, queue):
        data = (X, y, sample_weight) if sample_weight else (X, y)
        policy = self._get_policy(queue, *data)
        data = to_table(*_convert_to_supported(policy, *data))
        params = self._get_onedal_params(data[0])
        train_result = module.train(policy, params, *data)

        self._onedal_model = train_result.model

        if self.oob_score:
            if isinstance(self, ClassifierMixin):
                self.oob_score_ = from_table(train_result.oob_err_accuracy).item()
                self.oob_decision_function_ = from_table(
                    train_result.oob_err_decision_function
                )

            else:
                self.oob_score_ = from_table(train_result.oob_err_r2).item()
                self.oob_prediction_ = from_table(train_result.oob_err_prediction)

        return self

    def _create_model(self, module):
        # TODO:
        # upate error msg.
        raise NotImplementedError("Creating model is not supported.")

    def _predict(self, X, module, queue, hparams=None):
        _check_is_fitted(self)
        policy = self._get_policy(queue, X)

        model = self._onedal_model
        X = to_table(_convert_to_supported(policy, X))
        params = self._get_onedal_params(X)
        if hparams is not None and not hparams.is_default:
            result = module.infer(policy, params, hparams.backend, model, X)
        else:
            result = module.infer(policy, params, model, X)

        return from_table(result.responses)

    def _predict_proba(self, X, module, queue, hparams=None):
        _check_is_fitted(self)
        policy = self._get_policy(queue, X)
        X = to_table(_convert_to_supported(policy, X))
        params = self._get_onedal_params(X)
        params["infer_mode"] = "class_probabilities"

        model = self._onedal_model
        if hparams is not None and not hparams.is_default:
            result = module.infer(policy, params, hparams.backend, model, X)
        else:
            result = module.infer(policy, params, model, X)

        return from_table(result.probabilities)


class RandomForestClassifier(ClassifierMixin, BaseForest, metaclass=ABCMeta):
    def __init__(
        self,
        n_estimators=100,
        criterion="gini",
        max_depth=None,
        min_samples_split=2,
        min_samples_leaf=1,
        min_weight_fraction_leaf=0.0,
        max_features="sqrt",
        max_leaf_nodes=None,
        min_impurity_decrease=0.0,
        min_impurity_split=None,
        bootstrap=True,
        oob_score=False,
        random_state=None,
        warm_start=False,
        class_weight=None,
        ccp_alpha=0.0,
        max_samples=None,
        max_bins=256,
        min_bin_size=1,
        infer_mode="class_responses",
        splitter_mode="best",
        voting_mode="weighted",
        error_metric_mode="none",
        variable_importance_mode="none",
        algorithm="hist",
        **kwargs,
    ):
        super().__init__(
            n_estimators=n_estimators,
            criterion=criterion,
            max_depth=max_depth,
            min_samples_split=min_samples_split,
            min_samples_leaf=min_samples_leaf,
            min_weight_fraction_leaf=min_weight_fraction_leaf,
            max_features=max_features,
            max_leaf_nodes=max_leaf_nodes,
            min_impurity_decrease=min_impurity_decrease,
            min_impurity_split=min_impurity_split,
            bootstrap=bootstrap,
            oob_score=oob_score,
            random_state=random_state,
            warm_start=warm_start,
            class_weight=class_weight,
            ccp_alpha=ccp_alpha,
            max_samples=max_samples,
            max_bins=max_bins,
            min_bin_size=min_bin_size,
            infer_mode=infer_mode,
            splitter_mode=splitter_mode,
            voting_mode=voting_mode,
            error_metric_mode=error_metric_mode,
            variable_importance_mode=variable_importance_mode,
            algorithm=algorithm,
        )

    def fit(self, X, y, sample_weight=None, queue=None):
        return self._fit(
            X,
            y,
            sample_weight,
            self._get_backend("decision_forest", "classification", None),
            queue,
        )

    def predict(self, X, queue=None):
        hparams = get_hyperparameters("decision_forest", "infer")
        return super()._predict(
            X,
            self._get_backend("decision_forest", "classification", None),
            queue,
            hparams,
        )

    def predict_proba(self, X, queue=None):
        hparams = get_hyperparameters("decision_forest", "infer")

        return super()._predict_proba(
            X,
            self._get_backend("decision_forest", "classification", None),
            queue,
            hparams,
        )


class RandomForestRegressor(RegressorMixin, BaseForest, metaclass=ABCMeta):
    def __init__(
        self,
        n_estimators=100,
        criterion="squared_error",
        max_depth=None,
        min_samples_split=2,
        min_samples_leaf=1,
        min_weight_fraction_leaf=0.0,
        max_features=1.0,
        max_leaf_nodes=None,
        min_impurity_decrease=0.0,
        min_impurity_split=None,
        bootstrap=True,
        oob_score=False,
        random_state=None,
        warm_start=False,
        class_weight=None,
        ccp_alpha=0.0,
        max_samples=None,
        max_bins=256,
        min_bin_size=1,
        infer_mode="class_responses",
        splitter_mode="best",
        voting_mode="weighted",
        error_metric_mode="none",
        variable_importance_mode="none",
        algorithm="hist",
    ):
        super().__init__(
            n_estimators=n_estimators,
            criterion=criterion,
            max_depth=max_depth,
            min_samples_split=min_samples_split,
            min_samples_leaf=min_samples_leaf,
            min_weight_fraction_leaf=min_weight_fraction_leaf,
            max_features=max_features,
            max_leaf_nodes=max_leaf_nodes,
            min_impurity_decrease=min_impurity_decrease,
            min_impurity_split=min_impurity_split,
            bootstrap=bootstrap,
            oob_score=oob_score,
            random_state=random_state,
            warm_start=warm_start,
            class_weight=class_weight,
            ccp_alpha=ccp_alpha,
            max_samples=max_samples,
            max_bins=max_bins,
            min_bin_size=min_bin_size,
            infer_mode=infer_mode,
            splitter_mode=splitter_mode,
            voting_mode=voting_mode,
            error_metric_mode=error_metric_mode,
            variable_importance_mode=variable_importance_mode,
            algorithm=algorithm,
        )

    def fit(self, X, y, sample_weight=None, queue=None):
        if sample_weight is not None:
            if hasattr(sample_weight, "__array__"):
                sample_weight[sample_weight == 0.0] = 1.0
            sample_weight = [sample_weight]
        return super()._fit(
            X,
            y,
            sample_weight,
            self._get_backend("decision_forest", "regression", None),
            queue,
        )

    def predict(self, X, queue=None):
        return super()._predict(
            X, self._get_backend("decision_forest", "regression", None), queue
        )


class ExtraTreesClassifier(ClassifierMixin, BaseForest, metaclass=ABCMeta):
    def __init__(
        self,
        n_estimators=100,
        criterion="gini",
        max_depth=None,
        min_samples_split=2,
        min_samples_leaf=1,
        min_weight_fraction_leaf=0.0,
        max_features="sqrt",
        max_leaf_nodes=None,
        min_impurity_decrease=0.0,
        min_impurity_split=None,
        bootstrap=False,
        oob_score=False,
        random_state=None,
        warm_start=False,
        class_weight=None,
        ccp_alpha=0.0,
        max_samples=None,
        max_bins=256,
        min_bin_size=1,
        infer_mode="class_responses",
        splitter_mode="random",
        voting_mode="weighted",
        error_metric_mode="none",
        variable_importance_mode="none",
        algorithm="hist",
    ):
        super().__init__(
            n_estimators=n_estimators,
            criterion=criterion,
            max_depth=max_depth,
            min_samples_split=min_samples_split,
            min_samples_leaf=min_samples_leaf,
            min_weight_fraction_leaf=min_weight_fraction_leaf,
            max_features=max_features,
            max_leaf_nodes=max_leaf_nodes,
            min_impurity_decrease=min_impurity_decrease,
            min_impurity_split=min_impurity_split,
            bootstrap=bootstrap,
            oob_score=oob_score,
            random_state=random_state,
            warm_start=warm_start,
            class_weight=class_weight,
            ccp_alpha=ccp_alpha,
            max_samples=max_samples,
            max_bins=max_bins,
            min_bin_size=min_bin_size,
            infer_mode=infer_mode,
            splitter_mode=splitter_mode,
            voting_mode=voting_mode,
            error_metric_mode=error_metric_mode,
            variable_importance_mode=variable_importance_mode,
            algorithm=algorithm,
        )

    def fit(self, X, y, sample_weight=None, queue=None):
        return self._fit(
            X,
            y,
            sample_weight,
            self._get_backend("decision_forest", "classification", None),
            queue,
        )

    def predict(self, X, queue=None):
        return super()._predict(
            X, self._get_backend("decision_forest", "classification", None), queue
        )

    def predict_proba(self, X, queue=None):
        return super()._predict_proba(
            X, self._get_backend("decision_forest", "classification", None), queue
        )


class ExtraTreesRegressor(RegressorMixin, BaseForest, metaclass=ABCMeta):
    def __init__(
        self,
        n_estimators=100,
        criterion="squared_error",
        max_depth=None,
        min_samples_split=2,
        min_samples_leaf=1,
        min_weight_fraction_leaf=0.0,
        max_features=1.0,
        max_leaf_nodes=None,
        min_impurity_decrease=0.0,
        min_impurity_split=None,
        bootstrap=False,
        oob_score=False,
        random_state=None,
        warm_start=False,
        class_weight=None,
        ccp_alpha=0.0,
        max_samples=None,
        max_bins=256,
        min_bin_size=1,
        infer_mode="class_responses",
        splitter_mode="random",
        voting_mode="weighted",
        error_metric_mode="none",
        variable_importance_mode="none",
        algorithm="hist",
    ):
        super().__init__(
            n_estimators=n_estimators,
            criterion=criterion,
            max_depth=max_depth,
            min_samples_split=min_samples_split,
            min_samples_leaf=min_samples_leaf,
            min_weight_fraction_leaf=min_weight_fraction_leaf,
            max_features=max_features,
            max_leaf_nodes=max_leaf_nodes,
            min_impurity_decrease=min_impurity_decrease,
            min_impurity_split=min_impurity_split,
            bootstrap=bootstrap,
            oob_score=oob_score,
            random_state=random_state,
            warm_start=warm_start,
            class_weight=class_weight,
            ccp_alpha=ccp_alpha,
            max_samples=max_samples,
            max_bins=max_bins,
            min_bin_size=min_bin_size,
            infer_mode=infer_mode,
            splitter_mode=splitter_mode,
            voting_mode=voting_mode,
            error_metric_mode=error_metric_mode,
            variable_importance_mode=variable_importance_mode,
            algorithm=algorithm,
        )

    def fit(self, X, y, sample_weight=None, queue=None):
        if sample_weight is not None:
            if hasattr(sample_weight, "__array__"):
                sample_weight[sample_weight == 0.0] = 1.0
            sample_weight = [sample_weight]
        return super()._fit(
            X,
            y,
            sample_weight,
            self._get_backend("decision_forest", "regression", None),
            queue,
        )

    def predict(self, X, queue=None):
        return super()._predict(
            X, self._get_backend("decision_forest", "regression", None), queue
        )

File: onedal/ensemble/forest.cpp
/*******************************************************************************
* Copyright 2023 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

#include "oneapi/dal/algo/decision_forest.hpp"

#include "onedal/common.hpp"
#include "onedal/version.hpp"

namespace py = pybind11;

namespace oneapi::dal::python {

template <typename Task, typename Ops>
struct method2t {
    method2t(const Task& task, const Ops& ops) : ops(ops) {}

    template <typename Float>
    auto operator()(const py::dict& params) {
        using namespace decision_forest;

        const auto method = params["method"].cast<std::string>();

        ONEDAL_PARAM_DISPATCH_VALUE(method, "dense", ops, Float, method::dense);
        ONEDAL_PARAM_DISPATCH_VALUE(method, "hist", ops, Float, method::hist);
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(method);
    }

    Ops ops;
};

std::vector<std::string> split(const std::string& str) {
    const auto size = str.size();
    std::vector<std::string> result;

    std::size_t start = 0;
    for (std::size_t i = 0; i < size; ++i) {
        if (str[i] == '|') {
            result.emplace_back(str.substr(start, i - start));
            start = i + 1;
        }
        else if (i == size - 1) {
            result.emplace_back(str.substr(start, i - start + 1));
        }
    }

    return result;
}

auto get_error_metric_mode(const py::dict& params) {
    using namespace decision_forest;

    auto mode = params["error_metric_mode"].cast<std::string>();
    auto modes = split(mode);
    const auto modes_num = modes.size();

    error_metric_mode result_mode = error_metric_mode::none;
    for (std::size_t i = 0; i < modes_num; ++i) {
        if (modes[i] == "none")
            result_mode |= error_metric_mode::none;
        else if (modes[i] == "out_of_bag_error")
            result_mode |= error_metric_mode::out_of_bag_error;
        else if (modes[i] == "out_of_bag_error_per_observation")
            result_mode |= error_metric_mode::out_of_bag_error_per_observation;
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20230101
        else if (modes[i] == "out_of_bag_error_accuracy")
            result_mode |= error_metric_mode::out_of_bag_error_accuracy;
        else if (modes[i] == "out_of_bag_error_r2")
            result_mode |= error_metric_mode::out_of_bag_error_r2;
        else if (modes[i] == "out_of_bag_error_decision_function")
            result_mode |= error_metric_mode::out_of_bag_error_decision_function;
        else if (modes[i] == "out_of_bag_error_prediction")
            result_mode |= error_metric_mode::out_of_bag_error_prediction;
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION>=20230101
        else
            ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(mode);
    }
    return result_mode;
}

auto get_infer_mode(const py::dict& params) {
    using namespace decision_forest;

    auto mode = params["infer_mode"].cast<std::string>();
    auto modes = split(mode);
    const auto modes_num = modes.size();

    infer_mode result_mode{};
    for (std::size_t i = 0; i < modes_num; ++i) {
        if (modes[i] == "class_responses")
            result_mode |= infer_mode::class_responses;
        else if (modes[i] == "class_probabilities")
            result_mode |= infer_mode::class_probabilities;
        else
            ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(mode);
    }
    return result_mode;
}

#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20230101
auto get_splitter_mode(const py::dict& params) {
    using namespace decision_forest;
    auto mode = params["splitter_mode"].cast<std::string>();
    if (mode == "best") {
        return splitter_mode::best;
    }
    else if (mode == "random") {
        return splitter_mode::random;
    }
    else
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(mode);
}
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION>=20230101

auto get_variable_importance_mode(const py::dict& params) {
    using namespace decision_forest;

    auto mode = params["variable_importance_mode"].cast<std::string>();
    if (mode == "none") {
        return variable_importance_mode::none;
    }
    else if (mode == "mdi") {
        return variable_importance_mode::mdi;
    }
    else if (mode == "mda_raw") {
        return variable_importance_mode::mda_raw;
    }
    else if (mode == "mda_scaled") {
        return variable_importance_mode::mda_scaled;
    }
    else
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(mode);
}

auto get_voting_mode(const py::dict& params) {
    using namespace decision_forest;

    auto mode = params["voting_mode"].cast<std::string>();
    if (mode == "weighted") {
        return voting_mode::weighted;
    }
    else if (mode == "unweighted") {
        return voting_mode::unweighted;
    }
    else
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(mode);
}

struct params2desc {
    template <typename Float, typename Method, typename Task>
    auto operator()(const pybind11::dict& params) {
        using namespace decision_forest;

        constexpr bool is_cls = std::is_same_v<Task, task::classification>;

        auto desc = descriptor<Float, Method, Task>{}
                        .set_observations_per_tree_fraction(
                            params["observations_per_tree_fraction"].cast<double>())
                        .set_impurity_threshold(params["impurity_threshold"].cast<double>())
                        .set_min_weight_fraction_in_leaf_node(
                            params["min_weight_fraction_in_leaf_node"].cast<double>())
                        .set_min_impurity_decrease_in_split_node(
                            params["min_impurity_decrease_in_split_node"].cast<double>())
                        .set_tree_count(params["tree_count"].cast<std::int64_t>())
                        .set_features_per_node(params["features_per_node"].cast<std::int64_t>())
                        .set_max_tree_depth(params["max_tree_depth"].cast<std::int64_t>())
                        .set_min_observations_in_leaf_node(
                            params["min_observations_in_leaf_node"].cast<std::int64_t>())
                        .set_min_observations_in_split_node(
                            params["min_observations_in_split_node"].cast<std::int64_t>())
                        .set_max_leaf_nodes(params["max_leaf_nodes"].cast<std::int64_t>())
                        .set_max_bins(params["max_bins"].cast<std::int64_t>())
                        .set_min_bin_size(params["min_bin_size"].cast<std::int64_t>())
                        .set_memory_saving_mode(params["memory_saving_mode"].cast<bool>())
                        .set_bootstrap(params["bootstrap"].cast<bool>())
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20230101
                        .set_splitter_mode(get_splitter_mode(params))
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20230101
                        .set_error_metric_mode(get_error_metric_mode(params))
                        .set_variable_importance_mode(get_variable_importance_mode(params));

        if constexpr (is_cls) {
            desc.set_class_count(params["class_count"].cast<std::int64_t>());
            desc.set_infer_mode(get_infer_mode(params));
            desc.set_voting_mode(get_voting_mode(params));
        }

#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000
        desc.set_seed(params["seed"].cast<std::int64_t>());
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240000

        return desc;
    }
};

template <typename Policy, typename Task>
void init_train_ops(py::module_& m) {
    m.def("train",
          [](const Policy& policy,
             const py::dict& params,
             const table& data,
             const table& responses,
             const table& weights) {
              using namespace decision_forest;
              using input_t = train_input<Task>;

              train_ops ops(policy, input_t{ data, responses, weights }, params2desc{});
              return fptype2t{ method2t{ Task{}, ops } }(params);
          });
    m.def("train",
          [](const Policy& policy,
             const py::dict& params,
             const table& data,
             const table& responses) {
              using namespace decision_forest;
              using input_t = train_input<Task>;

              train_ops ops(policy, input_t{ data, responses }, params2desc{});
              return fptype2t{ method2t{ Task{}, ops } }(params);
          });
}

template <typename Policy, typename Task>
void init_infer_ops(py::module_& m) {
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240300
    using infer_hyperparams_t = decision_forest::detail::infer_parameters<Task>;
    m.def("infer",
          [](const Policy& policy,
             const py::dict& params,
             const infer_hyperparams_t& hyperparams,
             const decision_forest::model<Task>& model,
             const table& data) {
              using namespace decision_forest;
              using input_t = infer_input<Task>;

              infer_ops_with_hyperparams ops(policy,
                                             input_t{ model, data },
                                             params2desc{},
                                             hyperparams);
              return fptype2t{ method2t{ Task{}, ops } }(params);
          });
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240300

    m.def("infer",
          [](const Policy& policy,
             const py::dict& params,
             const decision_forest::model<Task>& model,
             const table& data) {
              using namespace decision_forest;
              using input_t = infer_input<Task>;

              infer_ops ops(policy, input_t{ model, data }, params2desc{});
              return fptype2t{ method2t{ Task{}, ops } }(params);
          });
}

template <typename Task>
void init_model(py::module_& m) {
    using namespace decision_forest;
    using model_t = model<Task>;

    auto cls = py::class_<model_t>(m, "model")
                   .def(py::init())
                   .def(py::pickle(
                       [](const model_t& m) {
                           return serialize(m);
                       },
                       [](const py::bytes& bytes) {
                           return deserialize<model_t>(bytes);
                       }))
                   .def_property_readonly("tree_count", &model_t::get_tree_count);

    constexpr bool is_classification = std::is_same_v<Task, task::classification>;

    if constexpr (is_classification) {
        auto class_count_getter = &model_t::template get_class_count<>;
        cls.def_property_readonly("class_count", class_count_getter);
    }
}

template <typename Task>
void init_train_result(py::module_& m) {
    using namespace decision_forest;
    using result_t = train_result<Task>;

    py::class_<result_t>(m, "train_result")
        .def(py::init())
        .DEF_ONEDAL_PY_PROPERTY(model, result_t)
        .DEF_ONEDAL_PY_PROPERTY(oob_err, result_t)
        .DEF_ONEDAL_PY_PROPERTY(oob_err_per_observation, result_t)
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20230101
        .DEF_ONEDAL_PY_PROPERTY(oob_err_accuracy, result_t)
        .DEF_ONEDAL_PY_PROPERTY(oob_err_r2, result_t)
        .DEF_ONEDAL_PY_PROPERTY(oob_err_decision_function, result_t)
        .DEF_ONEDAL_PY_PROPERTY(oob_err_prediction, result_t)
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION>=20230101
        .DEF_ONEDAL_PY_PROPERTY(var_importance, result_t);
}

template <typename Task>
void init_infer_result(py::module_& m) {
    using namespace decision_forest;
    using result_t = infer_result<Task>;

    auto cls = py::class_<result_t>(m, "infer_result")
                   .def(py::init())
                   .DEF_ONEDAL_PY_PROPERTY(responses, result_t);

    constexpr bool is_classification = std::is_same_v<Task, task::classification>;

    if constexpr (is_classification) {
        // workaround for gcc which cannot deduce setters directly passed to def_property()
        auto proba_getter = &result_t::template get_probabilities<>;
        auto proba_setter = &result_t::template set_probabilities<>;
        cls.def_property("probabilities", proba_getter, proba_setter);
    }
}

#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240300
template <typename Task>
void init_infer_hyperparameters(py::module_& m) {
    using namespace dal::decision_forest::detail;
    using infer_hyperparams_t = infer_parameters<Task>;

    auto cls = py::class_<infer_hyperparams_t>(m, "infer_hyperparameters")
                   .def(py::init())
                   .def("set_block_size",
                        [](infer_hyperparams_t& self, std::int64_t block_size) {
                            self.set_block_size(block_size);
                        })
                   .def("get_block_size",
                        [](const infer_hyperparams_t& self) -> std::int64_t {
                            return self.get_block_size();
                        })
                   .def("set_min_trees_for_threading",
                        [](infer_hyperparams_t& self, std::int64_t trees) {
                            self.set_min_trees_for_threading(trees);
                        })
                   .def("get_min_trees_for_threading",
                        [](const infer_hyperparams_t& self) -> std::int64_t {
                            return self.get_min_trees_for_threading();
                        })
                   .def("set_min_number_of_rows_for_vect_seq_compute",
                        [](infer_hyperparams_t& self, std::int64_t rows) {
                            self.set_min_number_of_rows_for_vect_seq_compute(rows);
                        })
                   .def("get_min_number_of_rows_for_vect_seq_compute",
                        [](const infer_hyperparams_t& self) -> std::int64_t {
                            return self.get_min_number_of_rows_for_vect_seq_compute();
                        })
                   .def("set_scale_factor_for_vect_parallel_compute",
                        [](infer_hyperparams_t& self, double factor) {
                            self.set_scale_factor_for_vect_parallel_compute(factor);
                        })
                   .def("get_scale_factor_for_vect_parallel_compute",
                        [](const infer_hyperparams_t& self) -> double {
                            return self.get_scale_factor_for_vect_parallel_compute();
                        });
}
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240300

ONEDAL_PY_TYPE2STR(decision_forest::task::classification, "classification");
ONEDAL_PY_TYPE2STR(decision_forest::task::regression, "regression");

ONEDAL_PY_DECLARE_INSTANTIATOR(init_model);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_train_result);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_infer_result);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_train_ops);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_infer_ops);
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240300
ONEDAL_PY_DECLARE_INSTANTIATOR(init_infer_hyperparameters);
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240300

ONEDAL_PY_INIT_MODULE(ensemble) {
    using namespace decision_forest;
    using namespace dal::detail;

    using task_list = types<task::classification, task::regression>;
    auto sub = m.def_submodule("decision_forest");

#ifdef ONEDAL_DATA_PARALLEL_SPMD
    ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_spmd, task_list);
    ONEDAL_PY_INSTANTIATE(init_infer_ops, sub, policy_spmd, task_list);
#else // ONEDAL_DATA_PARALLEL_SPMD
    ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_list, task_list);
    ONEDAL_PY_INSTANTIATE(init_infer_ops, sub, policy_list, task_list);

    ONEDAL_PY_INSTANTIATE(init_model, sub, task_list);
    ONEDAL_PY_INSTANTIATE(init_train_result, sub, task_list);
    ONEDAL_PY_INSTANTIATE(init_infer_result, sub, task_list);

#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240300
    ONEDAL_PY_INSTANTIATE(init_infer_hyperparameters, sub, task_list);
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240300

#endif // ONEDAL_DATA_PARALLEL_SPMD
}

} // namespace oneapi::dal::python
